File: js/src/Vista.SDK/Codebook.ts
import { CodebookName, CodebookNames, MetadataTag } from '.';
import { CodebookDto } from './types/CodebookDto';
import { isNullOrWhiteSpace, tryParseInt } from './util/util';

export class Codebook {
  public name: CodebookName;
  private readonly _groupMap: Map<string, string>;
  private readonly _standardValues: CodebookStandardValues;
  private readonly _groups: CodebookGroups;

  private static readonly tagAlphabet = 'abcdefghijklmnopqrstuvwxyz0123456789.';
  private static readonly positionTagAlphabet = this.tagAlphabet + '-';

  public constructor(dto: CodebookDto) {
    this.name = CodebookNames.fromString(dto.name);

    this._groupMap = new Map<string, string>();

    const data = Object.entries(dto.values).flatMap(([key, values]) =>
      values.map(v => ({ group: key, value: v.trim() }))
    );

    for (const t of data) {
      this._groupMap.set(t.value, t.group);
    }

    const valueSet = new Set<string>(data.map(t => t.value));
    const groupSet = new Set<string>(data.map(t => t.group));

    this._standardValues = new CodebookStandardValues(this.name, valueSet);
    this._groups = new CodebookGroups(this.name, groupSet);
  }

  public get groups() {
    return this._groups;
  }

  public get standardValues() {
    return this._standardValues;
  }

  public hasGroup(group: string): boolean {
    return this._groups.contains(group);
  }
  public hasStandardValue(value: string): boolean {
    return this._standardValues.contains(value);
  }

  public tryCreateTag(value?: string): MetadataTag | undefined {
    if (!value || isNullOrWhiteSpace(value)) return;

    let isCustom = false;
    if (this.name === CodebookName.Position) {
      if (this.validatePosition(value) < 100) return;
      if (this.validatePosition(value) === PositionValidationResult.Custom) {
        isCustom = true;
      }
    } else {
      if (![...value].every(c => Codebook.tagAlphabet.includes(c))) return;
      if (this.name !== CodebookName.Detail && !this._standardValues.contains(value)) {
        isCustom = true;
      }
    }
    return new MetadataTag(this.name, value, isCustom);
  }

  public createTag(value?: string): MetadataTag {
    const tag = this.tryCreateTag(value);
    if (!tag) throw new Error(`Invalid value for metadata tag: codebook=${this.name}, value=${value}`);
    return tag;
  }

  public validatePosition(position: string): PositionValidationResult {
    if (isNullOrWhiteSpace(position)) return PositionValidationResult.Invalid;

    if (position.trim().length != position.length) return PositionValidationResult.Invalid;

    if (![...position].every(c => Codebook.positionTagAlphabet.includes(c))) return PositionValidationResult.Invalid;

    if (this._standardValues.contains(position)) return PositionValidationResult.Valid;

    if (typeof tryParseInt(position) === 'number') return PositionValidationResult.Valid;

    if (!position.includes('-')) return PositionValidationResult.Custom;

    const positions = position.split('-');
    const validations: PositionValidationResult[] = [];
    for (const positionStr of positions) {
      validations.push(this.validatePosition(positionStr));
    }

    if (!!validations.find(v => v < 100)) return Math.max(...validations);

    const numberNotAtEnd = !!positions.find((pValue, pIndex) => tryParseInt(pValue) && pIndex < positions.length - 1);
    const positionsWithoutNumber = positions.filter(p => !tryParseInt(p));
    const alphabeticallySorted = positionsWithoutNumber.sort();
    const notAlphabeticallySorted = JSON.stringify(positionsWithoutNumber) !== JSON.stringify(alphabeticallySorted);

    if (numberNotAtEnd || notAlphabeticallySorted) return PositionValidationResult.InvalidOrder;

    if (validations.every(v => v === PositionValidationResult.Valid)) {
      const groups = positions.map(p => (tryParseInt(p) ? '<number>' : this._groupMap.get(p) ?? ''));

      const groupsSet = new Set<string>(groups);
      if (!groups.includes('DEFAULT_GROUP') && groupsSet.size !== groups.length)
        return PositionValidationResult.InvalidGrouping;
    }

    return Math.max(...validations);
  }
}

export enum PositionValidationResult {
  Invalid = 0,
  InvalidOrder = 1,
  InvalidGrouping = 2,
  Valid = 100,
  Custom = 101,
}

class CodebookStandardValues {
  private readonly _name: CodebookName;
  private readonly _standardValues: Set<string>;

  public constructor(name: CodebookName, standardValues: Set<string>) {
    this._name = name;
    this._standardValues = standardValues;
  }

  public get values() {
    return this._standardValues;
  }

  public contains(tagValue: string): boolean {
    if (this._name === CodebookName.Position && tryParseInt(tagValue)) return true;
    return this._standardValues.has(tagValue);
  }
}

class CodebookGroups {
  private readonly _name: CodebookName;
  private readonly _groups: Set<string>;

  public constructor(name: CodebookName, groups: Set<string>) {
    this._name = name;
    this._groups = groups;
  }

  public get values() {
    return this._groups;
  }

  public get count() {
    return this._groups.size;
  }

  public contains(group: string): boolean {
    return this._groups.has(group);
  }
}

File: csharp/src/Vista.SDK/Codebook.cs
namespace Vista.SDK;

public sealed record class Codebook
{
    public CodebookName Name { get; private init; }
    private readonly Dictionary<string, string> _groupMap;
    private readonly CodebookStandardValues _standardValues;
    private readonly CodebookGroups _groups;

    private const string TagAlphabet = "abcdefghijklmnopqrstuvwxyz0123456789.";
    private const string PositionTagAlphabet = TagAlphabet + "-";

    internal Codebook(CodebookDto dto)
    {
        Name = dto.Name switch
        {
            "positions" => CodebookName.Position,
            "calculations" => CodebookName.Calculation,
            "quantities" => CodebookName.Quantity,
            "states" => CodebookName.State,
            "contents" => CodebookName.Content,
            "commands" => CodebookName.Command,
            "types" => CodebookName.Type,
            "detail" => CodebookName.Detail,
            _ => throw new ArgumentException("Unknown metadata tag: " + dto.Name, nameof(dto.Name)),
        };

        _groupMap = new();

        var data = dto.Values
            .SelectMany(kvp => kvp.Value.Select(v => (Group: kvp.Key.Trim(), Value: v.Trim())))
            .Where(v => v.Value != "<number>")
            .ToArray();

        foreach (var t in data)
            _groupMap[t.Value] = t.Group;

        var valueSet = new HashSet<string>(data.Select(t => t.Value));
        var groupSet = new HashSet<string>(data.Select(t => t.Group));
        _standardValues = new CodebookStandardValues(Name, valueSet);
        _groups = new CodebookGroups(Name, groupSet);
    }

    public CodebookGroups Groups => _groups;

    public CodebookStandardValues StandardValues => _standardValues;

    public bool HasGroup(string group) => _groups.Contains(group);

    public bool HasStandardValue(string value) => _standardValues.Contains(value);

    public MetadataTag? TryCreateTag(string? value)
    {
        if (value is null || string.IsNullOrWhiteSpace(value))
            return null;

        var isCustom = false;

        if (Name.Equals(CodebookName.Position))
        {
            if ((int)ValidatePosition(value) < 100)
                return null;

            if (ValidatePosition(value).Equals(PositionValidationResult.Custom))
                isCustom = true;
        }
        else
        {
            if (!value.All(c => TagAlphabet.Contains(c)))
                return null;

            if (Name != CodebookName.Detail && !StandardValues.Contains(value))
                isCustom = true;
        }

        return new MetadataTag(Name, value, isCustom);
    }

    public MetadataTag CreateTag(string value)
    {
        var tag = TryCreateTag(value);
        if (tag is null)
            throw new ArgumentException(
                $"Invalid value for metadata tag: codebook={Name}, value={value}"
            );

        return tag.Value;
    }

    public PositionValidationResult ValidatePosition(string position)
    {
        if (string.IsNullOrWhiteSpace(position))
            return PositionValidationResult.Invalid;

        if (position.Trim().Length != position.Length)
            return PositionValidationResult.Invalid;

        if (!position.All(c => PositionTagAlphabet.Contains(c)))
            return PositionValidationResult.Invalid;

        if (StandardValues.Contains(position))
            return PositionValidationResult.Valid;

        if (int.TryParse(position, out _))
            return PositionValidationResult.Valid;

        if (!position.Contains('-'))
            return PositionValidationResult.Custom;

        var positions = position.Split('-');
        var validations = new List<PositionValidationResult>();
        foreach (var positionStr in positions)
        {
            validations.Add(ValidatePosition(positionStr));
        }

        if (validations.Any(v => (int)v < 100))
            return validations.Max();

        var numberNotAtEnd = positions
            .Where(
                (pValue, pIndex) => int.TryParse(pValue, out _) && pIndex < (positions.Length - 1)
            )
            .Any();

        var positionsWithoutNumber = positions.Where(p => !int.TryParse(p, out _)).ToList();
        var alphabeticallySorted = positionsWithoutNumber.OrderBy(p => p).ToList();
        var notAlphabeticallySorted = !positionsWithoutNumber.SequenceEqual(alphabeticallySorted);

        if (numberNotAtEnd || notAlphabeticallySorted)
            return PositionValidationResult.InvalidOrder;

        if (validations.All(v => (int)v == (int)PositionValidationResult.Valid))
        {
            var groups = positions
                .Select(p => int.TryParse(p, out _) ? "<number>" : _groupMap[p])
                .ToArray();

            var groupsSet = new HashSet<string>(groups);
            if (!groups.Contains("DEFAULT_GROUP") && groupsSet.Count != groups.Length)
                return PositionValidationResult.InvalidGrouping;
        }

        return validations.Max();
    }
}

public enum PositionValidationResult
{
    Invalid = 0,
    InvalidOrder = 1,
    InvalidGrouping = 2,

    Valid = 100,
    Custom = 101
}

public sealed class CodebookStandardValues
{
    private readonly CodebookName _name;
    private readonly HashSet<string> _standardValues;

    internal CodebookStandardValues(CodebookName name, HashSet<string> standardValues)
    {
        _name = name;
        _standardValues = standardValues;
    }

    public bool Contains(string tagValue)
    {
        if (_name == CodebookName.Position && int.TryParse(tagValue, out _))
            return true;

        return _standardValues.Contains(tagValue);
    }
}

public sealed class CodebookGroups
{
    private readonly CodebookName _name;
    private readonly HashSet<string> _groups;

    internal CodebookGroups(CodebookName name, HashSet<string> groups)
    {
        _name = name;
        _groups = groups;
    }

    public int Count => _groups.Count;

    public bool Contains(string group) => _groups.Contains(group);
}
